### ISSUE TITLE
Future of CRF layer

### ISSUE THREAD

**Issue by ivyleavedtoadflax** [wellcometrust/deep_reference_parser#28]
The CRF causes some problems, namely:

*  dependency on `keras_contribs` which:
  * Is not compatible (afaik) with tensorflow 2.0 (which is where the latest versions of keras live)
  * Prevents deep_reference_parser being published on pypi (it doesn't allow static dependencies)
* Requires logic to rebuild the model prior to predictions in the same way that the model is built for training, rather than just loading the complete model from a file.

Replacing the CRF with an attention layer (or similar) would ameliorate these problems. Note that it is already possible to remove it right now by specifying `output="softmax"` rather than `"crf"` when building the model with `deep_reference_parser.build_model()`. A softmax output will almost certainly perform worse than a CRF though.

**Comment by ivyleavedtoadflax** [comment]
Looks like CRF is available in tf 2.0 https://www.tensorflow.org/addons/api_docs/python/tfa/text/crf

**Comment by ivyleavedtoadflax** [comment]
Here's an example, but note that the CRF module is now in tfa.text.crf, not contrib

**Comment by ivyleavedtoadflax** [comment]
Ahh it is implemented for tf but not for tf.keras, though looks like it could be coming: https://github.com/tensorflow/addons/pull/377#pullrequestreview-335963486

**Comment by ivyleavedtoadflax** [comment]
This has been merged: https://github.com/tensorflow/addons/pull/1999

**Comment by chaalic** [comment]
Hello, when using CRF layer with BI-LSTM for an NER task, i get the following error : 
crf_loss  *
crf, idx = y_pred._keras_history[:2]

    AttributeError: 'Tensor' object has no attribute '_keras_history'
I get that it's a problem in the loss function, but I don't know how to get past it. Could you please help if you have found a solution ?

**Comment by ivyleavedtoadflax** [comment]
Hi @chaalic is this code you are running outside of the deep reference parser?

**Comment by chaalic** [comment]
Yes it's for another task of named entity recognition, but the model i'm using is the same : bilstm with CRF.

**Comment by ivyleavedtoadflax** [comment]
Ah OK. If you post some more of your code here we may be able to spot something.
